### ISSUE TITLE
F.one_hot to return BoolTensor/ByteTensor instead of LongTensor?

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#33046]
What is the reason for F.one_hot to return a rather large int64 type intead of int8 types such as Bool or Byte?
Its values are always 0/1 and with existing type promotion, manipulation with Bool/Byte in cross-entropy type formulas should be ok.

Related issue: https://github.com/pytorch/pytorch/issues/15457
Related PR about specifying dtype in one_hot call: https://github.com/pytorch/pytorch/pull/29672. The difference of this issue is about changing the default dtype.

cc @albanD @mruberry

**Comment by fmassa** [comment]
What's the primary use-case for `one_hot`? If it's to pass it in `cross_entropy`, then having it be of `float` dtype would be more useful. So if we were to change the default (which I'm not sure it's worth it), I'd maybe change it to float instead.

But before anything, I think that adding a `dtype` argument would be a first step before anything, if we want to change the default.

**Comment by vadimkantorov** [comment]
If cross_entropy functions now work transparently with type promotion, then even bool would work, but I agree: first dtype argument is a good right step!

**Comment by vadimkantorov** [comment]
(side note about my usecase: I use F.one_hot to construct some binary masks and later combine them with masks from other sources, so Bool outputs make more sense in this context)

**Comment by Linux-cpp-lisp** [comment]
For what its worth, in our applications (ML for atomic systems), it is very common to use a one-hot encoding of atom species (element type, such as carbon, oxygen, etc.) as an initial input encoding for our models. Currently we have to use `.to(torch.double)` after `one_hot` to get floating point feature vectors we can use, so it would be great if `one_hot` also supported outputting floating point dtypes directly as well.

**Comment by vadimkantorov** [comment]
Could default be changed to torch.bool? Or would that be BC-incompat?

**Comment by jbschlosser** [comment]
> Could default be changed to torch.bool? Or would that be BC-incompat?

Yeah, changing the default would unfortunately be BC-breaking.

**Comment by vadimkantorov** [comment]
@jbschlosser Could you please reopen this? The PR https://github.com/pytorch/pytorch/pull/58090 seems reverted

**Comment by XuehaiPan** [comment]
I've encountered an issue where `F.one_hot` does not support `dtype` keyword argument.

For consistency and flexibility, would it be possible to enhance `F.one_hot()` to accept the full range of `dtype`s that `torch.ones()` supports? Additionally, extending `dtype` support for `F.one_hot()` would help minimize type conversions and the creation of unnecessary intermediate tensor objects.

```python
end_indices = ...
ret = F.one_hot(end_indices, num_classes=sequence_length).float()
```
